### PR TITLE
EZP-29721: Broken URL alias history is not restored after regenerating aliases

### PR DESCRIPTION
**Pull request by alongosz**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-29721](https://jira.ez.no/browse/EZP-29721), follow-up to [EZP-29139](https://jira.ez.no/browse/EZP-29139)
| **Bug/Improvement**| yes
| **New feature**    | no
| **Target version** | `5.4`, `6.7`+
| **BC breaks**      | not yet
| **Tests pass**     | no
| **Doc needed**     | no

This is an attempt to fix edge case discovered by @mnocon - when parent of a nested URL alias was gone due to data corruption, the command displayed error referring to itself. It was fixed by moving cleanup prior to regenerating aliases, which was rather a workaround.

## Solution

This solution consists of several parts:
1. Cleaning corrupted aliases is done after performing refresh (21a5726).
2. `URLAliasService::refreshSystemUrlAliasesForLocation` performs cleanup of `link` and `parent` Ids based on the original (current) URL alias entry (99b2da5).
3. The method `UrlAlias\Handler::archiveUrlAliasesForDeletedTranslations` which caused the error mentioned earlier was refactored and optimized to avoid building path at this point as it is not necessary for the method to work (13a8a70). Doing cleanup of what causes an issue here for each Location would be a performance bottleneck. It's done globally later on by calling `URLAliasService::deleteCorruptedUrlAliases`.
4. There is an integration test showing the use case (3209535).

**TODO**:
- [x] Restore broken history.
- [x] Implement tests.
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

### CODE REVIEW

File: eZ/Publish/API/Repository/Tests/BaseTest.php [hunk 1]
```diff
@@ -555,6 +556,60 @@ public function createUserWithPolicies($login, array $policiesData)
...
+     */
+    protected function getRawDatabaseConnection()
+    {
+        $connection = $this
+            ->getSetupFactory()
+            ->getServiceContainer()->get('ezpublish.api.storage_engine.legacy.connection');
+
+        if (!$connection instanceof Connection) {
+            throw new \RuntimeException(
+                sprintf('Expected %s got something else', Connection::class)
```
Comment (mikadamczyk) [inline eZ/Publish/API/Repository/Tests/BaseTest.php]: Can we improve `something else` part?

File: eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Gateway/DoctrineDatabase.php [hunk 2]
```diff
@@ -1310,6 +1310,93 @@ public function deleteUrlAliasesWithBrokenLink()
...
+            ->update('ezurlalias_ml')
+            ->set('link', ':linkId')
+            ->set('parent', ':parentId')
+            ->where(
+                $expr->eq('action', ':action')
+            )
+            ->andWhere(
+                $expr->eq(
+                    'is_original',
+                    $updateQueryBuilder->createNamedParameter(0, \PDO::PARAM_INT)
```
Comment (mikadamczyk) [inline eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Gateway/DoctrineDatabase.php]: Should not we use `ParameterType::INTEGER`?

File: eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Mapper.php [hunk 3]
```diff
@@ -75,6 +75,26 @@ public function extractUrlAliasListFromData(array $rows)
...
+     *
+     * @param array $rows
+     *
+     * @return array
+     */
+    public function extractLanguageCodesFromData(array $rows)
+    {
+        $languageCodes = [];
+        foreach ($rows as $row) {
+            $languageCodes = array_merge(
```
Comment (mikadamczyk) [inline eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Mapper.php]: array merge in foreach is greedy construction, can we avoid it?

File: eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Gateway/DoctrineDatabase.php (hunk 2 above)
Comment (alongosz) [reply eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Gateway/DoctrineDatabase.php]: > Should not we use `ParameterType::INTEGER`?

On newer branches yes, on 1.7 there might be older Doctrine used AFAIR.
